README.md
<!-- Copyright © 2018–2023 Trevor Spiteri -->

<!-- Copying and distribution of this file, with or without
modification, are permitted in any medium without royalty provided the
copyright notice and this notice are preserved. This file is offered
as-is, without any warranty. -->

# Fixed-point numbers

The [*fixed* crate] provides fixed-point numbers.

  * [`FixedI8`] and [`FixedU8`] are eight-bit fixed-point numbers.
  * [`FixedI16`] and [`FixedU16`] are 16-bit fixed-point numbers.
  * [`FixedI32`] and [`FixedU32`] are 32-bit fixed-point numbers.
  * [`FixedI64`] and [`FixedU64`] are 64-bit fixed-point numbers.
  * [`FixedI128`] and [`FixedU128`] are 128-bit fixed-point numbers.

An <i>n</i>-bit fixed-point number has <i>f</i>&nbsp;=&nbsp;`Frac` fractional
bits where 0&nbsp;≤&nbsp;<i>f</i>&nbsp;≤&nbsp;<i>n</i>, and
<i>n</i>&nbsp;&minus;&nbsp;<i>f</i> integer bits. For example,
<code>[FixedI32]\<[U24]></code> is a 32-bit signed fixed-point number with
<i>n</i>&nbsp;=&nbsp;32 total bits, <i>f</i>&nbsp;=&nbsp;24 fractional bits, and
<i>n</i>&nbsp;&minus;&nbsp;<i>f</i>&nbsp;=&nbsp;8 integer bits.
<code>[FixedI32]\<[U0]></code> behaves like [`i32`], and
<code>[FixedU32]\<[U0]></code> behaves like [`u32`].

The difference between any two successive representable numbers is constant
throughout the possible range for a fixed-point number:
<i>Δ</i>&nbsp;=&nbsp;1/2<sup><i>f</i></sup>. When <i>f</i>&nbsp;=&nbsp;0, like
in <code>[FixedI32]\<[U0]></code>, <i>Δ</i>&nbsp;=&nbsp;1 because representable
numbers are integers, and the difference between two successive integers is 1.
When <i>f</i>&nbsp;=&nbsp;<i>n</i>, <i>Δ</i>&nbsp;=&nbsp;1/2<sup><i>n</i></sup>
and the value lies in the range &minus;0.5&nbsp;≤&nbsp;<i>x</i>&nbsp;<&nbsp;0.5
for signed numbers like <code>[FixedI32]\<[U32]></code>, and in the range
0&nbsp;≤&nbsp;<i>x</i>&nbsp;<&nbsp;1 for unsigned numbers like
<code>[FixedU32]\<[U32]></code>.

In version 1 the [*typenum* crate] is used for the fractional bit count `Frac`;
the plan is to to have a major version 2 with const generics when the Rust
compiler’s [`generic_const_exprs` feature] is ready and stabilized. An [alpha
version] is already available.

[`generic_const_exprs` feature]: https://github.com/rust-lang/rust/issues/76560
[alpha version]: https://docs.rs/fixed/2.0.0-alpha/fixed/

The main features are

  * Representation of binary fixed-point numbers up to 128 bits wide.
  * Conversions between fixed-point numbers and numeric primitives.
  * Comparisons between fixed-point numbers and numeric primitives.
  * Parsing from strings in decimal, binary, octal and hexadecimal.
  * Display as decimal, binary, octal and hexadecimal.
  * Arithmetic and logic operations.

This crate does *not* provide decimal fixed-point numbers. For example 0.001
cannot be represented exactly, as it is 1/10<sup>3</sup>. It is binary fractions
like 1/2<sup>4</sup> (0.0625) that can be represented exactly, provided there
are enough fractional bits.

This crate does *not* provide general analytic functions.

  * No algebraic functions are provided, for example no `sqrt` or `pow`.
  * No trigonometric functions are provided, for example no `sin` or `cos`.
  * No other transcendental functions are provided, for example no `log` or
    `exp`.

These functions are not provided because different implementations can have
different trade-offs, for example trading some correctness for speed.
Implementations can be provided in other crates.

  * The [*fixed-sqrt* crate] provides the square root operation.
  * The [*cordic* crate] provides various functions implemented using the
    [CORDIC] algorithm.

The conversions supported cover the following cases.

  * Infallible lossless conversions between fixed-point numbers and numeric
    primitives are provided using [`From`] and [`Into`]. These never fail
    (infallible) and do not lose any bits (lossless).
  * Infallible lossy conversions between fixed-point numbers and numeric
    primitives are provided using the [`LossyFrom`] and [`LossyInto`] traits.
    The source can have more fractional bits than the destination.
  * Checked lossless conversions between fixed-point numbers and numeric
    primitives are provided using the [`LosslessTryFrom`] and
    [`LosslessTryInto`] traits. The source cannot have more fractional bits than
    the destination.
  * Checked conversions between fixed-point numbers and numeric primitives are
    provided using the [`FromFixed`] and [`ToFixed`] traits, or using the
    [`from_num`] and [`to_num`] methods and [their checked
    versions][`checked_from_num`].
  * Additionally, [`az`] casts are implemented for conversion between
    fixed-point numbers and numeric primitives.
  * Fixed-point numbers can be parsed from decimal strings using [`FromStr`],
    and from binary, octal and hexadecimal strings using the
    [`from_str_binary`], [`from_str_octal`] and [`from_str_hex`] methods. The
    result is rounded to the nearest, with ties rounded to even.
  * Fixed-point numbers can be converted to strings using [`Display`],
    [`Binary`], [`Octal`], [`LowerHex`], [`UpperHex`], [`LowerExp`] and
    [`UpperExp`]. The output is rounded to the nearest, with ties rounded to
    even.
  * All fixed-point numbers are plain old data, so [`bytemuck`] bit casting
    conversions can be used.

## What’s new

### Version 1.23.0 news (2023-02-06)

  * String parsing of binary, octal and hexadecimal strings now supports base-2
    exponents starting with the separator “`p`” or “`P`”.
  * Bug fix: [`int_log10`][f-il10-1-23], [`int_log`][f-il-1-23],
    [`checked_int_log10`][f-cil10-1-23] and [`checked_int_log`][f-cil-1-23] were
    panicking or returning incorrect values for fixed-point numbers with no
    integer bits.

[f-cil-1-23]: https://docs.rs/fixed/~1.23/fixed/struct.FixedI32.html#method.checked_int_log
[f-cil10-1-23]: https://docs.rs/fixed/~1.23/fixed/struct.FixedI32.html#method.checked_int_log10
[f-il-1-23]: https://docs.rs/fixed/~1.23/fixed/struct.FixedI32.html#method.int_log
[f-il10-1-23]: https://docs.rs/fixed/~1.23/fixed/struct.FixedI32.html#method.int_log10

### Version 1.22.1 news (2023-01-26)

  * The errors from the [`lit`][f-l-1-22] method have been improved.
  * The `track_caller` attribute is now applied to many more inline functions
    that can panic.

### Version 1.22.0 news (2023-01-25)

  * String parsing now supports an optional exponent for all supported radices.
  * The [`lit`][f-l-1-22] method was added to all fixed-point numbers. This is
    useful to write fixed-point numbers literally in code and also works in
    constant context.
  * All fixed-point numbers now implement [`LowerExp`] and [`UpperExp`].
  * The [`Fixed`][tf-1-22] and [`FixedBits`][fb-1-22] traits now have the added
    supertraits [`Binary`], [`Octal`], [`LowerHex`], [`UpperHex`], [`LowerExp`]
    and [`UpperExp`].
  * The [`Wrapping`][w-1-22] and [`Unwrapped`][u-1-22] wrappers now implement
    [`Binary`], [`Octal`], [`LowerHex`], [`UpperHex`], [`LowerExp`] and
    [`UpperExp`].
  * The implementation of [`Debug`] for fixed-point numbers now has more relaxed
    constraints on the fractional bits generic parameter: now `Frac` only needs
    to implement [`Unsigned`][uns-1-22].

[`Debug`]: https://doc.rust-lang.org/nightly/core/fmt/trait.Debug.html
[f-l-1-22]: https://docs.rs/fixed/~1.22/fixed/struct.FixedI32.html#method.lit
[fb-1-22]: https://docs.rs/fixed/~1.22/fixed/traits/trait.FixedBits.html
[tf-1-22]: https://docs.rs/fixed/~1.22/fixed/traits/trait.Fixed.html
[u-1-22]: https://docs.rs/fixed/~1.22/fixed/struct.Unwrapped.html
[uns-1-22]: https://docs.rs/fixed/~1.22/fixed/types/extra/trait.Unsigned.html
[w-1-22]: https://docs.rs/fixed/~1.22/fixed/struct.Wrapping.html

### Other releases

Details on other releases can be found in [*RELEASES.md*].

[*RELEASES.md*]: https://gitlab.com/tspiteri/fixed/blob/master/RELEASES.md

## Quick examples

```rust
use fixed::types::I20F12;

// 19/3 = 6 1/3
let six_and_third = I20F12::from_num(19) / 3;
// four decimal digits for 12 binary digits
assert_eq!(six_and_third.to_string(), "6.3333");
// find the ceil and convert to i32
assert_eq!(six_and_third.ceil().to_num::<i32>(), 7);
// we can also compare directly to integers
assert_eq!(six_and_third.ceil(), 7);
```

The type [`I20F12`] is a 32-bit fixed-point signed number with 20 integer bits
and 12 fractional bits. It is an alias to <code>[FixedI32]\<[U12]></code>. The
unsigned counterpart would be [`U20F12`]. Aliases are provided for all
combinations of integer and fractional bits adding up to a total of eight, 16,
32, 64 or 128 bits.

```rust
use fixed::types::{I4F4, I4F12};

// -8 ≤ I4F4 < 8 with steps of 1/16 (~0.06)
let a = I4F4::from_num(1);
// multiplication and division by integers are possible
let ans1 = a / 5 * 17;
// 1 / 5 × 17 = 3 2/5 (3.4), but we get 3 3/16 (~3.2)
assert_eq!(ans1, I4F4::from_bits((3 << 4) + 3));
assert_eq!(ans1.to_string(), "3.2");

// -8 ≤ I4F12 < 8 with steps of 1/4096 (~0.0002)
let wider_a = I4F12::from(a);
let wider_ans = wider_a / 5 * 17;
let ans2 = I4F4::from_num(wider_ans);
// now the answer is the much closer 3 6/16 (~3.4)
assert_eq!(ans2, I4F4::from_bits((3 << 4) + 6));
assert_eq!(ans2.to_string(), "3.4");
```

The second example shows some precision and conversion issues. The low precision
of `a` means that `a / 5` is 3⁄16 instead of 1⁄5, leading to an inaccurate
result `ans1` = 3 3⁄16 (~3.2). With a higher precision, we get `wider_a / 5`
equal to 819⁄4096, leading to a more accurate intermediate result `wider_ans` =
3 1635⁄4096. When we convert back to four fractional bits, we get `ans2` = 3
6⁄16 (~3.4).

Note that we can convert from [`I4F4`] to [`I4F12`] using [`From`], as the
target type has the same number of integer bits and a larger number of
fractional bits. Converting from [`I4F12`] to [`I4F4`] cannot use [`From`] as we
have less fractional bits, so we use [`from_num`] instead.

## Writing fixed-point constants and values literally

The [`lit`] method, which is available as a `const` function, can be used to
parse literals. It supports
  * underscores as separators;
  * prefixes “`0b`”, “`0o`” and “`0x`” for binary, octal and hexadecimal
    numbers;
  * an optional decimal exponent with separator “`e`” or “`E`” for decimal,
    binary and octal numbers, or with separator “`@`” for all supported radices
    including hexadecimal.

```rust
use fixed::types::I16F16;

// 0.1275e2 is 12.75
const TWELVE_POINT_75: I16F16 = I16F16::lit("0.127_5e2");
// 1.8 hexadecimal is 1.5 decimal, and 18@-1 is 1.8
const ONE_POINT_5: I16F16 = I16F16::lit("0x_18@-1");
// 12.75 + 1.5 = 14.25
let sum = TWELVE_POINT_75 + ONE_POINT_5;
assert_eq!(sum, 14.25);
```

The [*fixed-macro* crate] is an alternative which provides a convenient macro to
write down fixed-point constants literally in the code. It supports underscores
as separators, binary/octal/hexadecimal integers, and an optional exponent for
decimal numbers, but it does not support fractions or exponents for
binary/octal/hexadecimal as they cannot be parsed by the Rust compiler.

```rust
use fixed::types::I16F16;
use fixed_macro::fixed;
use fixed_macro::types::I16F16;

// 0.1275e2 is 12.75
const TWELVE_POINT_75: I16F16 = fixed!(0.127_5e2: I16F16);
// 11 binary is 3 decimal
const THREE: I16F16 = I16F16!(0b_11);
// 12.75 + 3 = 15.75
let sum = TWELVE_POINT_75 + THREE;
assert_eq!(sum, 15.75);
```

## Using the *fixed* crate

The *fixed* crate is available on [crates.io][*fixed* crate]. To use it in your
crate, add it as a dependency inside [*Cargo.toml*]:

```toml
[dependencies]
fixed = "1.23"
```

The *fixed* crate requires rustc version 1.61.0 or later.

## Optional features

The *fixed* crate has these optional feature:

 1. `arbitrary`, disabled by default. This provides the generation of arbitrary
    fixed-point numbers from raw, unstructured data. This feature requires the
    [*arbitrary* crate].
 2. `serde`, disabled by default. This provides serialization support for the
    fixed-point types. This feature requires the [*serde* crate].
 3. `std`, disabled by default. This is for features that are not possible under
    `no_std`: currently the implementation of the [`Error`] trait for
    [`ParseFixedError`].
 4. `serde-str`, disabled by default. Fixed-point numbers are serialized as
    strings showing the value when using human-readable formats. This feature
    requires the `serde` and the `std` optional features. **Warning:** numbers
    serialized when this feature is enabled cannot be deserialized when this
    feature is disabled, and vice versa.

To enable features, you can add the dependency like this to [*Cargo.toml*]:

```toml
[dependencies.fixed]
version = "1.23"
features = ["serde"]
```

## Experimental optional features

It is not considered a breaking change if the following experimental features
are removed. The removal of experimental features would however require a minor
version bump. Similarly, on a minor version bump, optional dependencies can be
updated to an incompatible newer version.

RELEASES.md
<!-- Copyright © 2018–2021 Trevor Spiteri -->

<!-- Copying and distribution of this file, with or without
modification, are permitted in any medium without royalty provided the
copyright notice and this notice are preserved. This file is offered
as-is, without any warranty. -->


Version 1.11.0-borsh0_10-mango (2023-09-13)
===========================
[`Implements`][issue 1] borsh 0.10 update to the last version without the bug https://gitlab.com/tspiteri/fixed/-/issues/57

[issue 1]: https://github.com/blockworks-foundation/fixed/issues/1

Version 1.11.0 (2021-11-24)
===========================

  * The following methods were added to all fixed-point numbers, to the
    [`Fixed`][tf-1-11] trait, and to the [`Wrapping`][w-1-11] and
    [`Unwrapped`][u-1-11] wrappers:
      * [`lerp`][f-l-1-11]
      * [`inv_lerp`][f-il-1-11]
  * The following methods were added to all fixed-point numbers and to the
    [`Fixed`][tf-1-11] trait:
      * [`checked_lerp`][f-cl-1-11], [`saturating_lerp`][f-sl-1-11],
        [`wrapping_lerp`][f-wl-1-11], [`unwrapped_lerp`][f-ul-1-11],
        [`overflowing_lerp`][f-ol-1-11]
      * [`checked_inv_lerp`][f-cil-1-11], [`saturating_inv_lerp`][f-sil-1-11],
        [`wrapping_inv_lerp`][f-wil-1-11], [`unwrapped_inv_lerp`][f-uil-1-11],
        [`overflowing_inv_lerp`][f-oil-1-11]
  * The [*typenum* crate] dependency was updated to [version
    1.14][typenum-1-14].
  * The [`LeEqU8`][leu8-1-11], [`LeEqU16`][leu16-1-11], [`LeEqU32`][leu32-1-11],
    [`LeEqU64`][leu64-1-11] and [`LeEqU128`][leu128-1-11] traits no longer have
    a direct `'static` constraint, as it is a constraint of their supertrait
    [`Unsigned`][uns-1-11] since typenum [version 1.14][typenum-1-14]. This
    fixes a potential compatibility issue introduced in version 1.9.0.
  * An experimental feature was added to enable serialization using the [*borsh*
    crate] ([merge request 9]).

[f-cil-1-11]: https://docs.rs/fixed/~1.11/fixed/struct.FixedI32.html#method.checked_inv_lerp
[f-cl-1-11]: https://docs.rs/fixed/~1.11/fixed/struct.FixedI32.html#method.checked_lerp
[f-il-1-11]: https://docs.rs/fixed/~1.11/fixed/struct.FixedI32.html#method.inv_lerp
[f-l-1-11]: https://docs.rs/fixed/~1.11/fixed/struct.FixedI32.html#method.lerp
[f-oil-1-11]: https://docs.rs/fixed/~1.11/fixed/struct.FixedI32.html#method.overflowing_inv_lerp
[f-ol-1-11]: https://docs.rs/fixed/~1.11/fixed/struct.FixedI32.html#method.overflowing_lerp
[f-sil-1-11]: https://docs.rs/fixed/~1.11/fixed/struct.FixedI32.html#method.saturating_inv_lerp
[f-sl-1-11]: https://docs.rs/fixed/~1.11/fixed/struct.FixedI32.html#method.saturating_lerp
[f-uil-1-11]: https://docs.rs/fixed/~1.11/fixed/struct.FixedI32.html#method.unwrapped_inv_lerp
[f-ul-1-11]: https://docs.rs/fixed/~1.11/fixed/struct.FixedI32.html#method.unwrapped_lerp
[f-wil-1-11]: https://docs.rs/fixed/~1.11/fixed/struct.FixedI32.html#method.wrapping_inv_lerp
[f-wl-1-11]: https://docs.rs/fixed/~1.11/fixed/struct.FixedI32.html#method.wrapping_lerp
[leu128-1-11]: https://docs.rs/fixed/~1.11/fixed/types/extra/trait.LeEqU128.html
[leu16-1-11]: https://docs.rs/fixed/~1.11/fixed/types/extra/trait.LeEqU16.html
[leu32-1-11]: https://docs.rs/fixed/~1.11/fixed/types/extra/trait.LeEqU32.html
[leu64-1-11]: https://docs.rs/fixed/~1.11/fixed/types/extra/trait.LeEqU64.html
[leu8-1-11]: https://docs.rs/fixed/~1.11/fixed/types/extra/trait.LeEqU8.html
[merge request 9]: https://gitlab.com/tspiteri/fixed/-/merge_requests/9
[tf-1-11]: https://docs.rs/fixed/~1.11/fixed/traits/trait.Fixed.html
[typenum-1-14]: https://docs.rs/typenum/~1.14/typenum/index.html
[u-1-11]: https://docs.rs/fixed/~1.11/fixed/struct.Unwrapped.html
[uns-1-11]: https://docs.rs/fixed/~1.11/fixed/types/extra/trait.Unsigned.html
[w-1-11]: https://docs.rs/fixed/~1.11/fixed/struct.Wrapping.html

Version 1.10.0 (2021-08-23)
===========================

  * The crate now requires rustc version 1.53.0 or later.
  * <code>{[Div][`Div`],[DivAssign][`DivAssign`],[Rem][`Rem`],[RemAssign][`RemAssign`]}\<[NonZeroU32][`NonZeroU32`]></code>
    are now implemented for [`FixedU32`][fu-1-10], and similar for all other
    unsigned fixed-point numbers.
  * <code>{[Rem][`Rem`],[RemAssign][`RemAssign`]}\<[NonZeroI32][`NonZeroI32`]></code>
    are now implemented for [`FixedI32`][fi-1-10], and similar for all other
    signed fixed-point numbers.
  * The new [`arbitrary`][feat-1-10] optional feature was added to implement the
    [`Arbitrary`][a-a-1] trait provided by the [*arbitrary* crate] for all
    fixed-point numbers ([issue 37]).

[a-a-1]: https://docs.rs/arbitrary/^1/arbitrary/trait.Arbitrary.html
[feat-1-10]: https://docs.rs/fixed/~1.10/fixed/index.html#optional-features
[fi-1-10]: https://docs.rs/fixed/~1.10/fixed/struct.FixedI32.html
[fu-1-10]: https://docs.rs/fixed/~1.10/fixed/struct.FixedU32.html
[issue 37]: https://gitlab.com/tspiteri/fixed/-/issues/37

Version 1.9.0 (2021-05-13)
==========================

  * Fixed-point numbers can now be formatted as hexadecimal with [`Debug`]
    similarly to primitive integers, for example formatting with `{:X?}` will
    produce upper-case hexadecimal fixed-point numbers.
  * The following methods were added to all fixed-point numbers, to the
    [`Fixed`][tf-1-9] trait, and to the [`Wrapping`][w-1-9] and
    [`Unwrapped`][u-1-9] wrappers:
      * [`is_zero`][f-iz-1-9]
      * [`dist`][f-d-1-9]
  * The following methods were added to all fixed-point numbers and to the
    [`Fixed`][tf-1-9] trait:
      * [`checked_dist`][f-cd-1-9], [`saturating_dist`][f-sd-1-9],
        [`wrapping_dist`][f-wd-1-9], [`unwrapped_dist`][f-ud-1-9],
        [`overflowing_dist`][f-od-1-9]
  * The [`unsigned_dist`][f-unsd-1-9] method was added to all signed
    fixed-point types and to the [`FixedSigned`][tfs-1-9] trait.
  * The following associated types and provided methods were added to the
    [`Fixed`][tf-1-9] trait:
      * [`Signed`][tf-s-1-9], [`Unsigned`][tf-u-1-9]

src/log10.rs
// Copyright © 2018–2023 Trevor Spiteri

// This library is free software: you can redistribute it and/or
// modify it under the terms of either
//
//   * the Apache License, Version 2.0 or
//   * the MIT License
//
// at your option.
//
// You should have recieved copies of the Apache License and the MIT
// License along with the library. If not, see
// <https://www.apache.org/licenses/LICENSE-2.0> and
// <https://opensource.org/licenses/MIT>.

// the value must be positive for all public functions

pub mod int_part {
    // 1 <= val <= MAX (255)
    // 0 <= log <= 2
    #[inline]
    pub const fn u8(val: u8) -> i32 {
        maybe_assert!(val >= 1);
        // | from |  to |  a |  b | a&b |
        // |    1 |   9 | 10 | 01 |  00 |
        // |   10 |  99 | 11 | 01 |  01 |
        // |  100 | MAX | 11 | 10 |  10 |
        let val = val as i32;
        let a = (0b11 << 8) - 10 + val;
        let b = (0b10 << 8) - 100 + val;
        (a & b) >> 8
    }

    // 1 <= val <= MAX (65_535)
    // 0 <= log <= 4
    #[inline]
    pub const fn u16(val: u16) -> i32 {
        maybe_assert!(val >= 1);
        less_than_5(val as u32)
    }

    // 1 <= val <= MAX
    // 0 <= log <= 9
    pub const fn u32(mut val: u32) -> i32 {
        maybe_assert!(val >= 1);
        let mut log = 0;
        if val >= 100_000 {
            val /= 100_000;
            log += 5;
        }
        log + less_than_5(val)
    }

    // 1 <= val <= MAX
    // 0 <= log <= 19
    pub const fn u64(mut val: u64) -> i32 {
        maybe_assert!(val >= 1);
        let mut log = 0;
        if val >= 10_000_000_000 {
            val /= 10_000_000_000;
            log += 10;
        }
        if val >= 100_000 {
            val /= 100_000;
            log += 5;
        }
        maybe_assert!(val <= u32::MAX as u64);
        log + less_than_5(val as u32)
    }

    // 1 <= val <= MAX
    // 0 <= log <= 38
    pub const fn u128(mut val: u128) -> i32 {
        maybe_assert!(val >= 1);
        let mut log = 0;
        if val >= 100_000_000_000_000_000_000_000_000_000_000 {
            val /= 100_000_000_000_000_000_000_000_000_000_000;
            maybe_assert!(val <= u32::MAX as u128);
            return 32 + u32(val as u32);
        }
        if val >= 10_000_000_000_000_000 {
            val /= 10_000_000_000_000_000;
            log += 16;
        }
        maybe_assert!(val <= u64::MAX as u128);
        log + u64(val as u64)
    }

    // 0 < val < 100_000
    // 0 <= log <= 4
    const fn less_than_5(val: u32) -> i32 {
        maybe_assert!(val < 100_000);
        // |  from |    to |   a |   b | a&b |   c |   d | c&d | a&b ^ c&d |
        // |     1 |     9 | 010 | 011 | 010 | 110 | 011 | 010 |       000 |
        // |    10 |    99 | 011 | 011 | 011 | 110 | 011 | 010 |       001 |
        // |   100 |   999 | 011 | 100 | 000 | 110 | 011 | 010 |       010 |
        // |  1000 |  9999 | 011 | 100 | 000 | 111 | 011 | 011 |       011 |
        // | 10000 | 99999 | 011 | 100 | 000 | 111 | 100 | 100 |       100 |
        let val = val as i32;
        let a = (0b011 << 17) - 10 + val;
        let b = (0b100 << 17) - 100 + val;
        let c = (0b111 << 17) - 1000 + val;
        let d = (0b100 << 17) - 10000 + val;
        ((a & b) ^ (c & d)) >> 17
    }
}

pub mod frac_part {
    // MAX / 1000 (0) < val <= MAX (255)
    // -3 <= log <= -1
    #[inline]
    pub const fn u8(val: u8) -> i32 {
        if val > 25 {
            -1
        } else if val > 2 {
            -2
        } else {
            maybe_assert!(val > 0);
            -3
        }
    }

    // MAX / 100_000 (0) < val <= MAX (65_535)
    // -5 <= log <= -1
    #[inline]
    pub const fn u16(val: u16) -> i32 {
        if val > 6553 {
            -1
        } else if val > 655 {
            -2
        } else if val > 65 {
            -3
        } else if val > 6 {
            -4
        } else {
            maybe_assert!(val > 0);
            -5
        }
    }

    // 0 < val <= MAX
    // -10 <= log <= -1
    pub const fn u32(mut val: u32) -> i32 {
        const MAX: u32 = u32::MAX;
        if val <= MAX / 100_000_000 {
            val *= 100_000_000;
            // At this point, we have shifted out 8 digits, and we can only shift out 2 more.
            // We can only check up to -2 more because -10 <= log <= -8.
            if val > MAX / 10 {
                -9
            } else {
                maybe_assert!(val > MAX / 100);
                -10
            }
        } else {
            greater_equal_m8_u32(val)
        }
    }

    // 0 < val <= MAX
    // -20 <= log <= -1
    pub const fn u64(mut val: u64) -> i32 {
        const MAX: u64 = u64::MAX;
        let mut log = 0;
        if val <= MAX / 10_000_000_000_000_000 {
            // After this, we can only check up to -4 more because -20 <= log <= -16.
            // That is, we can skip the checks against MAX / 100_000_000 and MAX / 10_000.
            val *= 10_000_000_000_000_000;
            log += -16;
        } else {
            if val <= MAX / 100_000_000 {
                val *= 100_000_000;
                log += -8;
            }
            if val <= MAX / 10_000 {
                val *= 10_000;
                log += -4;
            }
        }
        log + if val > MAX / 10 {
            -1
        } else if val > MAX / 100 {
            -2
        } else if val > MAX / 1000 {
            -3
        } else {
            maybe_assert!(val > MAX / 10_000);
            -4
        }
    }

    // 0 < val <= MAX
    // -39 <= log <= -1
    pub const fn u128(mut val: u128) -> i32 {
        const MAX: u128 = u128::MAX;
        let mut log = 0;
        if val <= MAX / 100_000_000_000_000_000_000_000_000_000_000 {
            val *= 100_000_000_000_000_000_000_000_000_000_000;
            // At this point we have shifted out 32 digits, and we can only shift out 7 more.
            // We can
            //   * use val >> 96 because we have shifted out 32 decimal digits (106 bits)
            //   * only check up to -8 more because -39 <= log <= -32
            return -32 + greater_equal_m8_u32((val >> 96) as u32);
        }
        if val <= MAX / 10_000_000_000_000_000 {
            val *= 10_000_000_000_000_000;
            log += -16;
        }
        if val <= MAX / 100_000_000 {
            val *= 100_000_000;
            log += -8;
        }
        if log == -24 {
            // At this point we have shifted out 24 digits, and we can only shift out 15 more.
            // We can
            //   * use val >> 64 because we have shifted out 24 decimal digits (79 bits)
            //   * only check up to -8 more because -32 <= log <= -24
            return -24 + greater_equal_m8_u64((val >> 64) as u64);
        }
        // We have *not* shifted out enough decimal digits, so we must *not* convert to u32 or u64.
        if val <= MAX / 10_000 {
            val *= 10_000;
            log += -4;
        }
        log + if val > MAX / 10 {
            -1
        } else if val > MAX / 100 {
            -2
        } else if val > MAX / 1000 {
            -3
        } else {
            maybe_assert!(val > MAX / 10_000);
            -4
        }
    }

    // MAX / 100_000_000 < val <= MAX
    // -8 <= log <= -1
    const fn greater_equal_m8_u32(mut val: u32) -> i32 {
        const MAX: u32 = u32::MAX;
        maybe_assert!(val > MAX / 100_000_000);
        let mut log = 0;
        if val <= MAX / 10_000 {
            val *= 10_000;
            log += -4;
        }
        log + if val > MAX / 10 {
            -1
        } else if val > MAX / 100 {
            -2
        } else if val > MAX / 1000 {
            -3
        } else {
            maybe_assert!(val > MAX / 10_000);
            -4
        }
    }

    // MAX / 100_000_000 < val <= MAX
    // -8 <= log <= 1
    const fn greater_equal_m8_u64(mut val: u64) -> i32 {
        const MAX: u64 = u64::MAX;
        maybe_assert!(val > MAX / 100_000_000);
        let mut log = 0;
        if val <= MAX / 10_000 {
            val *= 10_000;
            log += -4;
        }
        log + if val > MAX / 10 {
            -1
        } else if val > MAX / 100 {
            -2
        } else if val > MAX / 1000 {
            -3
        } else {
            maybe_assert!(val > MAX / 10_000);
            -4
        }
    }
}

// check log10() and log(10) in tandem
#[cfg(test)]
mod tests {
    use crate::{log, log10};

    macro_rules! check_loop {
        ($T:ident) => {
            for i in 0..=log10::int_part::$T(<$T>::MAX) {
                let p = (10 as $T).pow(i as u32);
                if i > 0 {
                    assert_eq!(log10::int_part::$T(p - 1), i - 1);
                    assert_eq!(log::int_part::$T(p - 1, 10), i - 1);
                }
                assert_eq!(log10::int_part::$T(p), i);
                assert_eq!(log::int_part::$T(p, 10), i);
                assert_eq!(log10::int_part::$T(p + 1), i);
                assert_eq!(log::int_part::$T(p + 1, 10), i);
            }

            for i in 0..-log10::frac_part::$T(1) {
                let p = <$T>::MAX / (10 as $T).pow(i as u32);
                if p > 1 {
                    assert_eq!(log10::frac_part::$T(p - 1), -1 - i);
                    assert_eq!(log::frac_part::$T(p - 1, 10), -1 - i);
                }
                assert_eq!(log10::frac_part::$T(p), -1 - i);
                assert_eq!(log::frac_part::$T(p, 10), -1 - i);
                if i > 0 {
                    assert_eq!(log10::frac_part::$T(p + 1), -i);
                    assert_eq!(log::frac_part::$T(p + 1, 10), -i);
                }
            }
        };
    }

    #[test]
    fn log10_u8() {
        assert_eq!(log10::int_part::u8(1), 0);
        assert_eq!(log::int_part::u8(1, 10), 0);
        assert_eq!(log10::int_part::u8(u8::MAX), 2);
        assert_eq!(log::int_part::u8(u8::MAX, 10), 2);
        assert_eq!(log10::frac_part::u8(1), -3);
        assert_eq!(log::frac_part::u8(1, 10), -3);
        assert_eq!(log10::frac_part::u8(u8::MAX), -1);
        assert_eq!(log::frac_part::u8(u8::MAX, 10), -1);

        check_loop! { u8 }
    }

    #[test]
    fn log10_u16() {
        assert_eq!(log10::int_part::u16(1), 0);
        assert_eq!(log::int_part::u16(1, 10), 0);
        assert_eq!(log10::int_part::u16(u16::MAX), 4);
        assert_eq!(log::int_part::u16(u16::MAX, 10), 4);
        assert_eq!(log10::frac_part::u16(1), -5);
        assert_eq!(log::frac_part::u16(1, 10), -5);
        assert_eq!(log10::frac_part::u16(u16::MAX), -1);
        assert_eq!(log::frac_part::u16(u16::MAX, 10), -1);

        check_loop! { u16 }
    }

    #[test]
    fn log10_u32() {
        assert_eq!(log10::int_part::u32(1), 0);
        assert_eq!(log::int_part::u32(1, 10), 0);
        assert_eq!(log10::int_part::u32(u32::MAX), 9);
        assert_eq!(log::int_part::u32(u32::MAX, 10), 9);
        assert_eq!(log10::frac_part::u32(1), -10);
        assert_eq!(log::frac_part::u32(1, 10), -10);
        assert_eq!(log10::frac_part::u32(u32::MAX), -1);
        assert_eq!(log::frac_part::u32(u32::MAX, 10), -1);

        check_loop! { u32 }
    }

    #[test]
    fn log10_u64() {
        assert_eq!(log10::int_part::u64(1), 0);
        assert_eq!(log::int_part::u64(1, 10), 0);
        assert_eq!(log10::int_part::u64(u64::MAX), 19);
        assert_eq!(log::int_part::u64(u64::MAX, 10), 19);
        assert_eq!(log10::frac_part::u64(1), -20);
        assert_eq!(log::frac_part::u64(1, 10), -20);
        assert_eq!(log10::frac_part::u64(u64::MAX), -1);
        assert_eq!(log::frac_part::u64(u64::MAX, 10), -1);

        check_loop! { u64 }
    }

    #[test]
    fn log10_u128() {
        assert_eq!(log10::int_part::u128(1), 0);
        assert_eq!(log::int_part::u128(1, 10), 0);
        assert_eq!(log10::int_part::u128(u128::MAX), 38);
        assert_eq!(log::int_part::u128(u128::MAX, 10), 38);
        assert_eq!(log10::frac_part::u128(1), -39);
        assert_eq!(log::frac_part::u128(1, 10), -39);
        assert_eq!(log10::frac_part::u128(u128::MAX), -1);
        assert_eq!(log::frac_part::u128(u128::MAX, 10), -1);

        check_loop! { u128 }
    }
}
